### ISSUE TITLE
Support latest ffmpeg decoder API

### ISSUE THREAD

**Issue by katajakasa** [katajakasa/SDL_kitchensink#33]
Currently SDL_kitchensink uses older (deprecated) decoder API from ffmpeg to make sure things work with earlier ffmpeg versions. However, sooner or later it's going to be deprecated, and the library needs to move to the new API.

Implement the new api beside the old one and support both for now.

**Comment by ebachard** [comment]
Hello,

I resync'ed your repository, and I saw the very good you did. Wow, congratulations : I tested all the binaries, and everything seems to work fine. Better, I even tested with the big bunny (nice crash test for player btw), and it works really well.

Just a little issue : when compiling, I saw they are some warnings due to deprecated API. The problem is I'm using ffmpeg 4.0, so maybe I'm a bit ahead ;-) 

And if I'm not wrong, this issue seems to be there to fix them. As a little contribution, I'll attach a diff. This little change fixes the av_frame_get_best_effort_timestamp warning. Feel free to add the change I propose in your code (or not).

Important:  I'll attach the diff as a .txt file, because I fixed zillions of DOS end lines in my fork, and it's not very safe to provide a PR in these conditions. Better do simple ;-)

Last but not least, if I can fix some other warnings, I'll attach a diff there.

Thanks again for your nice project !

The diff:

[best_effort_timestamp_ericb.diff.txt](https://github.com/katajakasa/SDL_kitchensink/files/2252564/best_effort_timestamp_ericb.diff.txt)

**Comment by katajakasa** [comment]
Hey and thanks :)

Patch looks good, slipped it in with 77d1376eb33c090379faaf6a27ebb82d269109e1

ffmpeg 4 does indeed seem to deprecate a lot of things. Looks like right now SDL_kitchensink compiles (even if there are deprecation warnings), but knowing ffmpeg API stability, it might not be the case for long, so I absolutely welcome any compatibility fixes :)

**Comment by ebachard** [comment]
In the same line, here are two other patches : 
- avcodec_decode_audio4_ericb.diff.txt : no more warnings, seems to work as expected
[avcodec_decode_audio4_ericb.diff.txt](https://github.com/katajakasa/SDL_kitchensink/files/2253557/avcodec_decode_audio4_ericb.diff.txt)
- avcodec_decode_video2_ericb.diff.txt : no more warnings, but seems to trigger some errors (maybe true errors)
[avcodec_decode_video2_ericb.diff.txt](https://github.com/katajakasa/SDL_kitchensink/files/2253569/avcodec_decode_video2_ericb.diff.txt)

Remarks on the video side : the changes seem to work well, but got some errors like described below. More strange, I tried to follow the recommandations (see the link in the diff) but the right implementation seems to be the opposite of the blog entry ? That's the reason why it took me a while before to obtain something working with the frames (no problem with sound though). Sorry, maybe I misunderstood the explanations and my bad english probably striked again ... 

````
$ ./complex test.mp4 
Source streams:
 * Stream #0: KIT_STREAMTYPE_VIDEO
 * Stream #1: KIT_STREAMTYPE_AUDIO
Media information:
 * Audio: ac3 (ATSC A/52A (AC-3)), threads=8, 48000Hz, 2ch, 2b, signed
 * Video: h264 (H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10), threads=8, 1920x1080
Duration: 10.000000 seconds
Texture type: SDL_PIXELFORMAT_YV12
Audio format: AUDIO_S16
Subtitle format: SDL_PIXELFORMAT_UNKNOWN
[swscaler @ 0x4214e40] bad src image pointers
[swscaler @ 0x4214e40] bad src image pointers
[swscaler @ 0x4214e40] bad src image pointers

````

As usual, you can integrate, or not, what you consider interesting (MIT license, or whatever similar).

-- 
ericb

**Comment by ebachard** [comment]
Reading one more time the instructions with the new API, I finally fixed the strange issues. The patch is attached. (please forget the previous one)

[avcodec_decode_video2_ericb_fixed.diff.txt](https://github.com/katajakasa/SDL_kitchensink/files/2256085/avcodec_decode_video2_ericb_fixed.diff.txt)

**Comment by ebachard** [comment]
Last part (maybe some issues with DOS end lines, please check)


All warnings are fixed for me (was the first step), and I think your project is not far to respect the new API. You really did a great work.

Attached:  the last codex / codec_context fixes.

[codec_warnings_fixed_ericb.diff.txt](https://github.com/katajakasa/SDL_kitchensink/files/2256914/codec_warnings_fixed_ericb.diff.txt)


Can you point me some directions to (eventually) help you to complete this issue ? Thanks in advance.

**Comment by katajakasa** [comment]
Yup, the patches seem valid to me. I'll take a better look and do some merges when I have a moment.

**Comment by katajakasa** [comment]
Merged #46 for compatibility with new codec_parameters API stuff and avcodec_parameters_to_context. Subtitle bug found & worked around. Now just the decoder functions left :)

**Comment by katajakasa** [comment]
Done. With PR #48 the last changes required for ffmpeg new api are now in master. Closing ticket as fixed -- new issues related to this should be opened as new tickets.
